### PR TITLE
Auto download next unread chapters while reading

### LINKED ISSUE #80: [Feature Request] "Download X chapters ahead"
### Why/User Benefit/User Problem
Right now you can do batch downloads of chapters, and sure, you can download like 50 chapters and that could cover you, but when you reach that 50th chapter you might lose internet connection or something, or you gotta go somewhere and you don't have any data. Anyway, my point is tachi could keep your downloads X number of chapters ahead, so you always have X amount of chapters downloaded ahead (would also help people with poor connection for whom streaming the chapters has a lot of buffering)

### What/Requirements
(explain how this feature would behave)
just a checkbox or something in the Download chapters menu that says "Keep downloading" or something clearer, or perhaps a global setting. or both (global setting overwritten manually)

### PR DESCRIPTION
**Pull request by nzoba**
This PR contains:

- Auto-download next *chosen_number* unread chapters while reading (activable via settings) after reading 20% of a chapter (Closes #80)
- New setting to auto-download new chapters only if manga is completely read (and same *chosen_number* than auto-download next unread)
- New library setting to only update completely read manga (taken from original tachiyomi, default false maybe should change for true ?)
- Add *(Downloaded)* text to chapter name in transition page if chapter is downloaded &rarr; ex: *Chapter 3 - I am the best chapter (Downloaded)*

| Download settings | Library settings | Downloaded text in transition |
|----|----|----|
| Download settings | Library settings | Downloaded in transition |

### CODE REVIEW

**Comment by Jays2Kings** [comment]
Haven't looked into the code yet but 
> New setting to auto-download new chapters only if manga is completely read (and same chosen_number than auto-download next unread)
I think I'll pass on this feature, even though it's been requested heavily, with the preloading of the next chapter in the reader already and the fact sources want use to cool it with all the traffic, and for sources that download slower than you can read (or just series with really short chapters), it's not worth taking this one in

> New library setting to only update completely read manga (taken from original tachiyomi, default false maybe should change for true ?)
Good addition, was going to take it anyway so this saves me time, not sure how up to date from tachi this is but make sure this has the "x series skipped" + learn more notification (I'll probably make more adjustments to it if arkon doesn't looking at the discord confusion)

> Add (Downloaded) text to chapter name in transition page if chapter is downloaded → ex: Chapter 3 - I am the best chapter (Downloaded)
Another good addition, especially with the main feature of this pr. however since I do not want the first feature it seems somewhat obnoxious now. 
Can you make some changes to it:
* Use ImageSpan and add the download icon at the end of the chapter text instead of "downloaded". If its too complex use Compound drawables at the start of the text view
* Only show download icon if the next (or previous) chapter is downloaded but the current chapter is not
* Likewise if the current chapter is downloaded, but the on transitioning to isn't, add a cloud icon and show that instead (This should help with someone not realizing theyre on a non downloaded chapter

**Comment by nzoba** [comment]
> New setting to auto-download new chapters only if manga is completely read (and same chosen_number than auto-download next unread)
> I think I'll pass on this feature, even though it's been requested heavily, with the preloading of the next chapter in the reader already and the fact sources want use to cool it with all the traffic, and for sources that download slower than you can read (or just series with really short chapters), it's not worth taking this one in

I guess you wanted to quote the first point. From my point of view, this feature should not increase the traffic. Since once a chapter is downloaded the source isn't called anymore. Or maybe I'm missing something but I don't see how pre-downloading would be worse than preloading for the sources (especially since the start of the next chapter would be preloaded even if you want to pause reading for some time while an already downloaded chapter is just kept in your phone storage and won't call the source when you start reading it again). Except for small chapters like you said but since with this option you can download a few chapters ahead, even if the download is slow I think it should be able to be fast enough to not do both preloading and downloading.

The first point may be quite controversial but the second point is especially in favor of less traffic since the auto-download of new chapters already exists, I just added an option to download only the chapters of completely read manga (to avoid unnecessary downloads and storage filled for nothing) and a max amount of chapters per manga

> New library setting to only update completely read manga (taken from original tachiyomi, default false maybe should change for true ?)
> Good addition, was going to take it anyway so this saves me time, not sure how up to date from tachi this is but make sure this has the "x series skipped" + learn more notification (I'll probably make more adjustments to it if arkon doesn't looking at the discord confusion)

I didn't add that but I'll look into it

> Add (Downloaded) text to chapter name in transition page if chapter is downloaded → ex: Chapter 3 - I am the best chapter (Downloaded)
> Another good addition, especially with the main feature of this pr. however since I do not want the first feature it seems somewhat obnoxious now.
> Can you make some changes to it:
> 
> * Use ImageSpan and add the download icon at the end of the chapter text instead of "downloaded". If its too complex use Compound drawables at the start of the text view
> * Only show download icon if the next (or previous) chapter is downloaded but the current chapter is not
> * Likewise if the current chapter is downloaded, but the on transitioning to isn't, add a cloud icon and show that instead (This should help with someone not realizing theyre on a non downloaded chapter

I agree, I tried quickly to add an image with the text but it was not as easy as I thought but you are right the text is quite bad I'll look into the icon solution again

**Comment by Jays2Kings** [comment]
> I guess you wanted to quote the first point. From my point of view, this feature should not increase the traffic. Since once a chapter is downloaded the source isn't called anymore. Or maybe I'm missing something but I don't see how pre-downloading would be worse than preloading for the sources (especially since the start of the next chapter would be preloaded even if you want to pause reading for some time while an already downloaded chapter is just kept in your phone storage and won't call the source when you start reading it again). Except for small chapters like you said but since with this option you can download a few chapters ahead, even if the download is slow I think it should be able to be fast enough to not do both preloading and downloading.

Looking at the code now, I don't think it'll do much for the issue, preloading or downloading at the same time in the previous chapter still will cause an issue with slower sources. If anything, changing it to a download would mean users would have to wait for chapter to finish downloading before they can move to the next, otherwise itll be a mix of fetching from the source while reading, and downloading the same chapter, which is what causes the double traffic.

Since streaming and reading caches the reader pages, and uses those for downloading instead of fetching the pages again, I have another proposal: For the reader activity, keep the preloader as is, but if the user exits the reader activity, just like the where the feature is for removing read downloads after reading, we can add there to continue downloading the next chapter(s). This way if someone wants to stream from the source after downloads they can and not have to wait for the chapter to be fully downloaded before continuing, if they dont, they can exit and not have to tap the download button.

> The first point may be quite controversial but the second point is especially in favor of less traffic since the auto-download of new chapters already exists, I just added an option to download only the chapters of completely read manga (to avoid unnecessary downloads and storage filled for nothing) and a max amount of chapters per manga

I just meant the first point, the updates to skipping global updates and downloading is a vital addition for this ddos machine

**Comment by nzoba** [comment]
I did a few things (4 last commits) :
1. Changed the "Downloaded" text in transition for an icon (visible only in the cases you asked for).
2. Added the skipped notification with learn more link. It's not exactly the same as main tachi since the `LibraryUpdateService` class is quite different but it works well, tell if you want some modifications (or do it yourself if you prefer)
3. Changed the auto-download while reading as you suggested to be run only when you quit the reader

**Review by Jays2Kings — COMMENTED** [review]
The comments have gotten seriously heavy in the end here so Id recommend another splitting of these prs:

1 PR for the skipped updates
1 PR for the download ahead/auto download (you can keep the image changes in that one but those changes should be made for compound drawable)

For the download ahead pr, you dont have to make the suggested changes for that, I'll base the suggestions off your pr like I did for langauge badges

**Comment by nzoba** [comment]
Damn, yeah that's a lot of comments lol
Yes you are right, I'll split it into 2 PRs later, too many features in a single PR is never a good idea (like my first huge PR, sorry again 🤯) either for the reviewer or the author.

**Comment by Jays2Kings** [comment]
Btw I'll likely take your changes + suggestions from this pr and manually add in the skipped manga logic. I'll leave the auto download stuff for you

**Comment by nzoba** [comment]
Sure, you can do it if you want
Sorry I didn't have time to work on the PRs these days, I'll deal with your requested changes today or during the weekend

**Comment by Jays2Kings** [comment]
> Sure, you can do it if you want Sorry I didn't have time to work on the PRs these days, I'll deal with your requested changes today or during the weekend

No need to rush, if anything you can focus on the download ahead feature, the skipping updates is easy enough to pull from your branch + upstream, and the download ahead can wait

**Comment by Jays2Kings** [comment]
So if you want to use this existing PR, you can strip out the skipped update and download restrictions logic, and keep only the download ahead stuff. Sorry for absolutely breaking this pr though...

**Comment by nzoba** [comment]
No worries I'll look into it :ok_hand: 
Actually it's better, so that this PR doesn't add/change too many different features

File: app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsDownloadController.kt
Comment (nzoba) [reply app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsDownloadController.kt:81]: I'm not too sure about all these changes tbh
1. I agree the setting `download only completely read` is not particularly useful most of the time except
   - For example I like to deactivate all the library restrictions to update all my library once in a while (like 2 times a month) 
   - When you update a particular manga via manga_details, you don't really want to download all the chapters if not necessary
2.  I'm not sure I understand the difference between the `Download chapters after reading` and `Download ahead if chapter is not downloaded`. Shouldn't one of them be a switch ?
3. I kind of like the setting to limit the number of downloaded chapters for new chapters (was the same as download while reading for now), some sources like to mass release chapters and I don't think you really want to download everything at once (especially to ease source traffic)

Comment (Jays2Kings) [reply app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsDownloadController.kt:81]: 1. The issue isnt really the setting itself but more the fact that what people always say: Tachiyomi has way too many options for a reading app. I get your use case for it but I dont think I want to add more complexity to the update/download restrictions just yet
2. I was trying to think of the wording of them, but it goes something like this
   * `Download ahead if chapter is not downloaded`: This is the switch option basically, if its never, the auto dowload is disabled. If it's on, depending on the number set it will start auto downloading. So if set the setting to 2, you finish reading chapter 3, and chapter 5 isn't downloaded it will download that chapter and more based on...
   * `Download chapters after reading` (this setting will be hidden if the above is off): This decides how many more chapters along with the first non-downloaded chapter to download. If this is set to 3 for example, it would download 5, 6, and 7. 
3. Same as point 1. I will say you should make a PR for point 1 and 2 to the main tachi repo and ill pull it back if they accept. It seems like a feature more for source damage control so it would make sense to ask them if they think the setting would make sense than me

Comment (nzoba) [reply app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsDownloadController.kt:81]: 1. Well I agree, it's far from necessary, I'll remove it
2. It's more clear but I don't really see the point of having 2 settings like that. As you said too many/complex settings is not good. Wouldn't it be enough to just set how many chapters you want downloaded ahead ?

If I remove the `auto-download only over wifi` and `download only completely read` settings and only keep (bold is what I add) :
- **Download after reading** ? `switch` (to avoid migration, as you said)
- Download new chapters ? `switch`
- **How many downloaded chapters you want per manga** (= auto-download restrictions, may need a better name) `intList`
- Categories for new chapters `triStateList`

Imo it seems simple enough while adding some useful features that would answer 95% of the cases. What do you think ?
We can even put all of that in an `Auto-download` category like main.

Comment (Jays2Kings) [reply app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsDownloadController.kt:81]: I kinda agree with point 2, but to a degree. I'd still say keep all the existing settings the same and add a single setting for download ahead/auto download and thats
"Download next unread chapters after reading": Never/Disabled, Next chapter, next 2, 3, 5
This we can use the logic to check the next _x_ chapter(s) and download them if they aren't (assuming the currently read chapter was downloaded. So yeah much more simple.

so just:
* Download new chapters
* Categories

* "Download ahead" (section could be without a title since its the one setting)
   * "Download next unread chapters after reading"

And like i said for previous point 3 maybe just make a pr to main for a download new chapters limit

Comment (nzoba) [reply app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsDownloadController.kt:81]: Yeah seems good like that ! I'll go with your first option since a single section could indeed be quite confusing
I'll implement/modify all that during the week
And alright I'll see to make a PR to main for the new chapters limit

**Comment by nzoba** [comment]
I think I'm done. Just a few things I've added:
- Compound drawables for the transition icons to the right of the texts. I think the rendering is better on the left, but with the compound, the previous/next and current chapter texts would no longer line up. If you have an idea, I (or you) can change it.
- An infoPreference below the setting to add some precision, let me know if you think of an improvement.
- A check that 20% (random value) of the chapter has been read to download after exiting the reader, to avoid auto-download by missed click on a chapter or just if you want to check the start of a chapter.

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt
Comment (nzoba) [reply app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt]: Just forgot to add this case, I'll look into it tomorrow

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt
Comment (Jays2Kings) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt:176]: i'd say 0.33 would be a better but its a random value all the same really 
also add a check if the currentChapters.previousChapter was downloaded and if it was or this chapter has progressed enough, then downloadNextChapters (also pass a bool stating if the last chapter was downloaded to downloadNextChapters)

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt
Comment (Jays2Kings) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt:514]: so with that parameter passed
```suggestion
        if (isChapterDownloaded || isPrevChapterDownloaded) {
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt
Comment (Jays2Kings) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt:515]: pass the chapter to this since we'll work backwards if only the previous chapter is downloaded
```suggestion
            val currentChapters = viewerChaptersRelay.value ?: return
            downloadAutoNextChapters(chaptersNumberToDownload, if (isChapterDownloaded) currentChapters.currChapter else currentChapters.prevChapter)
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt
Comment (Jays2Kings) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderPresenter.kt:527]:
```suggestion
    private fun downloadAutoNextChapters(choice: Int, chapter: ReaderChapter?) {
        val chaptersToDownload = getNextUnreadChaptersSorted(chapter).take(choice)
        if (chaptersToDownload.isNotEmpty()) {
            downloadChapters(chaptersToDownload)
        }
    }

    private fun getNextUnreadChaptersSorted(chapter: ReaderChapter?): List<ChapterItem> {
        val currentChapterId = (chapter ?: getCurrentChapter())?.chapter?.id
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt
Comment (Jays2Kings) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt:84]:
```suggestion
        manga: Manga,
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt
Comment (Jays2Kings) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt:47]:
```suggestion
        manga: Manga,
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt
Comment (Jays2Kings) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt:117]: use setCompoundDrawablesRelativeWithIntrinsicBounds instead for RTL locales

File: app/src/main/res/values/strings.xml
Comment (Jays2Kings) [inline app/src/main/res/values/strings.xml:939]:
```suggestion
    <string name="download_ahead_info">Download ahead only applies to entries in library and if the last read chapter was downloaded</string>
```

**Review by Jays2Kings — COMMENTED** [review]
Yeah this looks great right now. I'll probably make a release before merging this since I dont want to add any more strings for this release if it can be helped

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt
Comment (nzoba) [reply app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/ReaderTransitionView.kt]: Out of curiosity, why add a comma here ? Just a personal preference ?

**Comment by nzoba** [comment]
I changed the logic a bit to download if **any** chapter from the last reading session was downloaded (last commit)
When exiting if the current chapter was not downloaded, it will be downloaded (include in number of download ahead fixed by setting), if not only the following chapters will be

**Review by Jays2Kings — COMMENTED** [review]
ok i think im done with comments, all logic looks great to me 👍 

i might make some changes to the strings and chapter transition after merge but i wont ask for them since they might not work

File: app/src/main/res/values/strings.xml
Comment (Jays2Kings) [inline app/src/main/res/values/strings.xml:939]: Less wordy and while simplifies the logic more than what it actually does, it gets the point across
```suggestion
    <string name="download_ahead_info">Download ahead only applies to entries in library and if the last chapter read was downloaded</string>
```

**Comment by nzoba** [comment]
This should be good.
I didn't add the plurals, I'll let you do that.

**Review by Jays2Kings — APPROVED** [review]
🚀 
Though I'll merge this after the next release

**Comment by nzoba** [comment]
Just a rebase to fix conflicts, nothing has changed

**Pull request merged**
